core/src/main/scala/org/apache/spark/TaskEndReason.scala

```diff
@@ -212,8 +215,8 @@ case object TaskResultLost extends TaskFailedReason {
  * Task was killed intentionally and needs to be rescheduled.
  */
 @DeveloperApi
-case object TaskKilled extends TaskFailedReason {
-  override def toErrorString: String = "TaskKilled (killed intentionally)"
+case class TaskKilled(reason: String, override val shouldRetry: Boolean) extends TaskFailedReason {
```

Comment:
I think it's a little weird for `shouldRetry` to be part of this interface: whether a killed task should be retried is already handled by existing scheduling logic and it's a little confusing to be setting it on a per-task basis. Basically, it's confusing to me because this isn't the source-of-truth on whether we can/will re-try and its purpose here is therefore a bit unclear to me.

As discussed offline, I think that we may be able to update the logic in `TaskSchedulerImpl.handleFailedTask` (the only place which reads this field) in order to eliminate the need for this field in this class.